### ISSUE TITLE
Reconcile Android/iOS on Dart entrypoint defaults when creating engine

### ISSUE THREAD

**Issue by xster** [flutter/flutter#33097]
Make Android and iOS uniform in needing to specify and defaults of 

- The entrypoint function
- Library
- Bundle

**Comment by matthew-carroll** [comment]
We'll need to chat about the specifics of this. I'm not in favor of passing nulls to methods for defaults, which apparently iOS is doing. Also, I'm not convinced that parity between these 2 platforms should supersede what would otherwise be a better API surface. Let's keep in mind that we will soon be on many more platforms, and unless we're going to try to impose Android and iOS API surfaces onto those platforms, this may be a losing proposition in general.

**Comment by xster** [comment]
I'm not specifically rooting for specifically using null everywhere but some degree of defaulting via some other constructor would be nice.

**Comment by matthew-carroll** [comment]
I can only think of 2 common reasons for defaults:

1. A developer doesn't really know how something works, so a default allows the developer to move forward without a deeper understanding.

2. Something must be done many times and a default reduces the tediousness.

I don't think either of these situations exist here.

For new developers who don't really understand what these different API pieces mean, they're almost certainly using `FlutterActivity` or `FlutterFragment`, which have defaults. Those have defaults because they're on the edges of the system. Any developer who is starting up a `FlutterEngine` needs to understand what they're doing, so a default doesn't save anything on that front.

As far as repetition, there are maybe 2 or 3 FlutterEngines in a substantial app. There will only be 1 FlutterEngine in the average app. So there isn't any highly repetitive work here that needs to be shortened.

So I still don't think we should be adding API surface to codify defaults within the central components of the Android API. I think there is probably no meaningful upside, but with an expanding API surface there is always an opportunity to regret baking in defaults later.

At best, I'd like to see numerous requests for these defaults from real customers before we seriously consider adding them.

**Comment by xster** [comment]
It's a bit of both 1 and 2, plus a bit of 3. what meaningful choice is the developer really making here outside of defaults.

Since we'd want to push users to use FlutterEngines even in simple cases, they're not really on the edges of the system. We're taxing them with learning about assets and bundles but in the end, I'm not sure we're really letting them make use of this new knowledge. When would they choose a different asset or bundle location?

**Comment by xster** [comment]
Tentatively assigning Matt. I think you recently mentioned looking at this.

**Comment by matthew-carroll** [comment]
@xster can you define the definition of done for this ticket? I've done some work on the Android API, but I'm not sure what this ticket is really looking for.

**Comment by xster** [comment]
I can't really recall what was already done or not :S. I think we can say this one's done if Android and iOS both have defaults for starting VM entrypoints or neither have defaults.

**Comment by matthew-carroll** [comment]
By "default" do you mean a factory method that creates a `DartyEntrypoint` configured for `main` and the standard apple bundle path? If so, it exists on Android. I don't know if it exists on iOS. Should I assign this to Aaron?

**Comment by xster** [comment]
Oh ya, great! Defaults exist for both the intent builder and the dart entrypoint constructors. I think this is done on Android.

For iOS, minor but having special meaning for nil on runWithEntrypoint:(NSString*)entrypoint is a bit non-ideal. Perhaps we can have a [FlutterEngine runDefaultEntrypoint]. But @gaaclarke can decide when he's back.

**Comment by gaaclarke** [comment]
We're unable to put the module in the run arguments since it will adversely affect app startup time, please see comment: https://github.com/flutter/engine/pull/12370#discussion_r328369135

**Comment by matthew-carroll** [comment]
@gaaclarke based on your last comment, should this be closed as "won't do"? Or is there still work that needs to be done?

**Comment by gaaclarke** [comment]
Here is where I ended up with this issue:
- Android does not yet support loading different modules, so we can't reconcile anything there.
- Android was proposing putting modules next to entry point specification, that would hurt their ability to do warm up, so we shouldn't do that.
- I added convenience methods to help with the construction of engines on iOS, as per xster's request, to make it as easy as android.

I have no further planned work.
